text/0000-ancillary-package-module-standardization.md

```diff
@@ -0,0 +1,234 @@
+
+Lastly, due to efforts like FastBoot, users expect to be able to run
+their JavaScript code both in the browser and in Node.js.  We need a
+system that makes our packages portable, so that they can be tested and
+run in any environment.
+
+# Detailed Design
+
+This RFC proposes that all packages under the Ember.js umbrella be
+incrementally migrated to standard JavaScript (ES6) modules.  This
+format is trivially transpilable into whatever ultimate module system
+the end user desires, whether that's {named, anonymous} AMD, CommonJS,
+or
+no module system at all using something like Rollup's bundle builds.
+
+JavaScript modules have the benefit of being statically analyzable, so
+tools can determine the dependency graph at build time.  By normalizing
+Ember's dependencies into JavaScript modules, we make it much easier to
+eliminate code that is not actually needed in final builds.
+
+To facilitate a backwards-compatible migration, we will standardize on a
+directory layout for projects, as well as a directory layout for the
+final built output. All packages will be distributed via npm.
+
+To ensure standardized behavior across repositories, a reusable npm
+package `broccoli-module-alchemist` will be created to encapsulate the
+decisions of this RFC.
+
+The source code for broccoli-module-alchemist is available at
+<https://github.com/monegraph/broccoli-module-alchemist>.
+
+## Supported Authoring Formats
+
+Authors may write package source code in TypeScript, ES2015, or
+a mix of both. Because transpilation is done using the TypeScript
+compiler, which supports both, projects can incrementally migrate
+towards TypeScript, or use it only in portions where it makes sense.
+
+Whether using TypeScript or JavaScript, authors should always use
+standard ES2015 module syntax for both importing and exporting modules.
+
+## Target Formats
+
+The original source code is compiled from TypeScript or ES2015+ to ES5
+to ensure maximum compatibility across browser and Node.js versions.
+
+Out of the box, several different module formats are emitted. Output
+targets can be configured on a package-by-package basis. For example, a
+package that only targets Node.js may want to author in TypeScript and
+only emit CommonJS-compatible modules.
+
+### CommonJS
+
+For use by Node.js and tools like browserify/webpack.
+
+### JavaScript (ES2015/ES6) Modules
+
+The standard JavaScript module syntax specified in ES2015, this
+statically-analyzable format is intended to be used by tools like Rollup
+or Ember CLI, which can analyze an entire app and its dependencies and
+strip out unused code.
+
+For Ember's packages, this is the output target that Ember will consume.
+
+Note that only the import/export syntax remains in the output; 
+all other ES2015 syntax is transpiled into ES5. This makes it easy for
+tools like Rollup to consume without having to add a transpiler like
+Babel.
+
+### UMD
+
+UMD modules can be consumed as CommonJS, AMD, or by exporting a global.
```

Comment:
@chadhietala The UMD stuff here is for other, non-Ember consumers. My hope is that Ember can jump to consuming the ES6 modules version, and the responsibility for compiling into something loader.js-compatible happens there. If not, it would be straightforward to add an AMD build.